### PR TITLE
add self hosted implementation and integration tests

### PR DESCRIPTION
**Pull request by steven-xu-lf**
- Add a username/password handler that talks to the self-hosted APIServer.
- Added extra details in the error reporting for the self hosted just like the repo client (will also do that for the cloud API in another PR/story) 
- Add integration tests for it and categorized those tests into cloud vs selfhosted
- configured the pipeline to not run the self hosted integration tests just like the C# library
- Add some UT for it 
- Formatted Documents
- Skipped a UT that stubs the completable future method for now (will have to figure out how to do it in the future) 

Reference:
https://github.com/Laserfiche/lf-api-client-core-dotnet/tree/1.x/src/APIServer
https://github.com/Laserfiche/lf-api-client-core-dotnet/blob/1.x/src/HttpHandlers/UsernamePasswordHandler.cs

### CODE REVIEW

File: src/main/java/com/laserfiche/api/client/APIServer/SessionKeyInfo.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package com.laserfiche.api.client.APIServer;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+
+public class SessionKeyInfo {
```
Comment (jason-chan-lf) [inline src/main/java/com/laserfiche/api/client/APIServer/SessionKeyInfo.java]: should this class be in model folder instead?

File: src/main/java/com/laserfiche/api/client/APIServer/SessionKeyInfo.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    @JsonProperty("access_token")
+    private String accessToken = null;
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/APIServer/SessionKeyInfo.java]: In the C# counterpart, the class with the same name has three fields with different name. Is there any reason that we name them different here? (Also, any reason we have different number of fields?)

Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/APIServer/SessionKeyInfo.java]: the rest of the fields for the other classes in this repo i believe are following the camel case naming convention, so I thought I would keep it consistent for this class, but i can change it to match that of the C# lib if you want. As for the number of fields in the class, i believe this class also has 3 fields with a setter and getter for each field

Comment (lf-jason-jia) [reply src/main/java/com/laserfiche/api/client/APIServer/SessionKeyInfo.java]: Oh sorry. I was looking at an outdated version of the .Net counterpart.

File: src/main/java/com/laserfiche/api/client/APIServer/TokenClient.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import java.util.concurrent.CompletableFuture;
+
+public interface TokenClient {
+
+    /**
+     * @param repoId Repository name
+     * @param body   Request body that contains username, passward and grant type
+     * @return Create an access token successfuly.
+     */
+
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/APIServer/TokenClient.java]: Shall we remove the empty line between method signature and documentation?

File: src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                .header("Accept", "application/json")
+                .header("Content-Type", "application/x-www-form-urlencoded")
+                .field("grant_type", "password")
+                .field("username", body
+                        .getUsername()
+                        .toString())
+                .field("password", body
+                        .getPassword()
+                        .toString())
+                .asObjectAsync(SessionKeyInfo.class);
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java]: Would this work? If we deserialize it into `SessionKeyInfo` but it's actualluy a JSON corresponds to `ProblemDetails`, I think it would throw exception like not being able to fit the JSON into `SessionKeyInfo`.

**Review by lf-jason-jia — CHANGES_REQUESTED** [review]
See comments

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+     *
+     * @param repositoryId Repository name
+     * @param username     The username used with "password" grant type.
+     * @param password     The password used with "password" grant type.
+     * @param baseUrl      APIServer Base Url e.g. https://example.com/LFRepositoryAPI
+     * @param client       OPTIONAL
+     */
+    public UsernamePasswordHandler(String repositoryId, String username, String password, String baseUrl,
+            TokenClient client) {
+        _baseUrl = baseUrl;
```
Comment (jason-chan-lf) [inline src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java]: I tried running the tests with baseUrl like `https://example.com/LFRepositoryAPI` and with an ending slash like `https://example.com/LFRepositoryAPI/`. When there is an ending slash the `beforeSendAsync_NewToken_Success` doesn't seem to pass. Maybe the `UsernamePasswordHandler` should handle both cases by stripping ending slash from baseUrl?

File: src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        return future.thenApply(httpResponse -> {
+            if (httpResponse.getStatus() != 200) {
+                ProblemDetails problemDetails;
+                try {
+                    String jsonString = new JSONObject(httpResponse.getBody()).toString();
+                    problemDetails = objectMapper.readValue(jsonString, ProblemDetails.class);
+                } catch (JsonProcessingException e) {
+                    e.printStackTrace();
+                    return null;
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java]: Not sure if we should terminate execution here. Like, maybe we should just throw the exception and not to return a null?

Comment (lf-jason-jia) [reply src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java]: Maybe it's okay as is.

File: src/test/java/com/laserfiche/api/client/integration/BaseTest.java [hunk 7]
```diff
@@ -21,6 +28,10 @@ public static void setUp() {
                     .load();
             // Read env variable
             accessKeyBase64 = dotenv.get("ACCESS_KEY");
+            repoId = dotenv.get("REPOSITORY_ID");
+            username = dotenv.get("APISERVER_USERNAME");
+            password = dotenv.get("APISERVER_PASSWORD");
+            baseUrl = dotenv.get("APISERVER_REPOSITORY_API_BASE_URL");
             spKey = dotenv.get("SERVICE_PRINCIPAL_KEY");
         }
         accessKey = AccessKey.createFromBase64EncodedAccessKey(accessKeyBase64);
```
Comment (jason-chan-lf) [inline src/test/java/com/laserfiche/api/client/integration/BaseTest.java]: btw i tried running the tests when ACCESS_KEY environment variable doesn't exist since it isnt needed for self-hosted and this would throw an exception. Maybe not needed to call this line if `accessKeyBase64` is null?

File: src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+                else if (httpResponse.getStatus() == 404)
+                    throw new ApiException("Requested attribute key not found.", httpResponse.getStatus(),
+                            httpResponse.getStatusText(), headersMap, problemDetails);
+                else if (httpResponse.getStatus() == 429)
+                    throw new ApiException("Rate limit is reached.", httpResponse.getStatus(),
+                            httpResponse.getStatusText(), headersMap, problemDetails);
+                else
+                    throw new RuntimeException(httpResponse.getStatusText());
+            }
+            return httpResponse.getBody();
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java]: We will never reach this line?

File: src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java (hunk 4 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java]: ok so i just tried to match the latest version of the auto generated code in the java repo client lib and updated the integration test associated with this test case

File: src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java (hunk 8 above)
Comment (sxu1998) [reply src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java]: hmm strange how this one file isn't updated after i pushed the latest changes, but this line should no longer be here after my latest commit

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+import com.laserfiche.api.client.APIServer.SessionKeyInfo;
+import com.laserfiche.api.client.APIServer.TokenClient;
+import com.laserfiche.api.client.APIServer.TokenClientImpl;
+import com.laserfiche.api.client.model.CreateConnectionRequest;
+
+import java.util.concurrent.CompletableFuture;
+
+public class UsernamePasswordHandler implements HttpRequestHandler {
+    private String _accessToken;
```
Comment (samadpaydar3) [inline src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java]: would you please consider removing the leading _, as it is not a common pattern in java.

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    private final String GRANTTYPE = "password";
+    private String _repoId;
```
Comment (samadpaydar3) [inline src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java]: maybe its better to rename to repositoryId

File: src/main/java/com/laserfiche/api/client/model/OffsetDateTimeDeserializer.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+import org.threeten.bp.OffsetDateTime;
+
+import java.io.IOException;
+
+public class OffsetDateTimeDeserializer extends StdDeserializer<OffsetDateTime> {
+    public OffsetDateTimeDeserializer() {
+        this(null);
+    }
+
+    public OffsetDateTimeDeserializer(Class<?> vc) {
```
Comment (samadpaydar3) [inline src/main/java/com/laserfiche/api/client/model/OffsetDateTimeDeserializer.java]: would you please rename vc to something more meaningful?
I couldn't guess what it stands for.

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
+package com.laserfiche.api.client.httphandlers;
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java:1]: Minor question on the naming: is username password an OAuth flow? It looks similar to this OAuth flow: https://auth0.com/docs/get-started/authentication-and-authorization-flow/resource-owner-password-flow

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java (hunk 9 above)
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java]: Let's not to add underscore (under "Variables" section): https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java (hunk 10 above)
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java]: Same for all underscored variables

File: src/main/java/com/laserfiche/api/client/model/ApiException.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
+package com.laserfiche.api.client.model;
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/model/ApiException.java:1]: Nothing to change. But a reminder that since the repository client also has this class, maybe we should make sure two classes are identical and the repository client can just use this class instead (it should be easy to do, we could do this before release).

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java (hunk 12 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java:1]: yes

File: src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+                .builder()
+                .addModule(module)
+                .disable(MapperFeature.CAN_OVERRIDE_ACCESS_MODIFIERS)
+                .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
+                .enable(MapperFeature.ACCEPT_CASE_INSENSITIVE_ENUMS)
+                .build();
+    }
+
+    @Override
+    public CompletableFuture<SessionKeyInfo> createAccessToken(String repoId, CreateConnectionRequest body) {
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java]: Why do we call `CreateConnectionRequest` body?

File: src/main/java/com/laserfiche/api/client/model/OffsetDateTimeDeserializer.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+import com.fasterxml.jackson.core.JacksonException;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.TreeNode;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
+import org.threeten.bp.OffsetDateTime;
+
+import java.io.IOException;
+
+public class OffsetDateTimeDeserializer extends StdDeserializer<OffsetDateTime> {
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/model/OffsetDateTimeDeserializer.java]: Again, similar to `ApiException`, aa reminder that we could probably ensure this one the same as the one in repository client and let the repository client uses this class.

File: src/main/java/com/laserfiche/api/client/model/ProblemDetails.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
+package com.laserfiche.api.client.model;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.HashMap;
+import java.util.Map;
+
+
+public class ProblemDetails extends HashMap<String, Object> {
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/model/ProblemDetails.java:9]: Again, reminder:
1. Ensure the one in repository client is the same as this one (it's not, it has builder API and this one doesn't).
2. Repository client can probably use this class instead.

File: src/test/java/com/laserfiche/api/client/integration/BaseTest.java [hunk 17]
```diff
@@ -8,10 +8,17 @@
...
     protected static AccessKey accessKey;
-
+    protected static String repoId;
+    protected static String username;
+    protected static String password;
+    protected static String baseUrl;
     @BeforeAll
     public static void setUp() {
         spKey = System.getenv("SERVICE_PRINCIPAL_KEY");
+        repoId = System.getenv("REPOSITORY_ID");
```
Comment (lf-jason-jia) [inline src/test/java/com/laserfiche/api/client/integration/BaseTest.java:18]: Reminder: we probably also want to add these in the GitHub pipeline in the future.

File: src/test/java/com/laserfiche/api/client/integration/UsernamePasswordHandlerTest.java [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.params.provider.Arguments.arguments;
+
+@Tag("APIServer")
```
Comment (lf-jason-jia) [inline src/test/java/com/laserfiche/api/client/integration/UsernamePasswordHandlerTest.java]: So we have `Tag("Cloud")`, would something like `Tag("SelfHosted")` make more sense than `@Tag("APIServer")`?

File: src/test/java/com/laserfiche/api/client/unit/UsernamePasswordHandlerTest.java [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.params.provider.Arguments.arguments;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.*;
+
+public class UsernamePasswordHandlerTest extends BaseTest {
```
Comment (lf-jason-jia) [inline src/test/java/com/laserfiche/api/client/unit/UsernamePasswordHandlerTest.java]: Does this one need a Tag?

File: src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java (hunk 14 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java]: just keeping it consistent with one the C# and how it is named in the swagger.json file for selfhosted

File: src/main/java/com/laserfiche/api/client/model/OffsetDateTimeDeserializer.java (hunk 11 above)
Comment (lf-jason-jia) [reply src/main/java/com/laserfiche/api/client/model/OffsetDateTimeDeserializer.java]: Yeah, I think this is copied from the repository client project. We probably also need a better name there.

File: src/test/java/com/laserfiche/api/client/unit/UsernamePasswordHandlerTest.java (hunk 19 above)
Comment (steven-xu-lf) [reply src/test/java/com/laserfiche/api/client/unit/UsernamePasswordHandlerTest.java]: i think all the UT r supposed to run in the pipeline, at least from what i see in the C# pipeline so no need to categorize and i do not see them categorized either in the C# repo

File: src/main/java/com/laserfiche/api/client/APIServer/TokenClient.java [hunk 20]
```diff
@@ -0,0 +1,16 @@
+package com.laserfiche.api.client.APIServer;
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/APIServer/TokenClient.java]: Would it make more sense to call this something like `package com.laserfiche.api.client.selfhosted;`?

Also, Java package name should be all lower case.

File: src/test/java/com/laserfiche/api/client/unit/UsernamePasswordHandlerTest.java [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+        handler
+                .afterSendAsync(mockedResponse)
+                .thenApply((shouldRetry) -> {
+                    // Assert
+                    assertEquals(false, shouldRetry);
+                    return null;
+                });
+    }
+
+    private static Stream<Arguments> ResponseOtherThanUnauthorized() {
```
Comment (lf-jason-jia) [inline src/test/java/com/laserfiche/api/client/unit/UsernamePasswordHandlerTest.java]: In Java, method name starts with lowercase letter.

File: src/test/java/com/laserfiche/api/client/integration/BaseTest.java (hunk 17 above)
Comment (steven-xu-lf) [reply src/test/java/com/laserfiche/api/client/integration/BaseTest.java:18]: so i addressed this with Paolo in our mobbing session last friday, and he suggested that we should run these on tfs instead of github and that i should create a low priority story for it which i will get to later

File: src/main/java/com/laserfiche/api/client/model/ProblemDetails.java (hunk 16 above)
Comment (sxu1998) [reply src/main/java/com/laserfiche/api/client/model/ProblemDetails.java:9]: can you specify which method in the repo client is needed, since this class in the repo client is auto-generated (i.e. they contain the schema annotation, generated annotation, and all those extra methods outside of getters and setters, which i believe you did say that i can remove in the prior PR if im not mistaken)

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+     *
+     * @param repositoryId Repository name
+     * @param username     The username used with "password" grant type.
+     * @param password     The password used with "password" grant type.
+     * @param baseUrl      APIServer Base Url e.g. https://example.com/LFRepositoryAPI
+     * @param client       OPTIONAL
+     */
+    public UsernamePasswordHandler(String repositoryId, String username, String password, String baseUrl,
+            TokenClient client) {
+        this.baseUrl = baseUrl.endsWith("LFRepositoryAPI") ? baseUrl : baseUrl.substring(0, baseUrl.lastIndexOf("/"));
```
Comment (jason-chan-lf) [inline src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java]: better to not hardcode `"LFRepositoryAPI"` Would this work?
`baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.lastIndexOf("/")) : baseUrl;`

Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java]: well i was thinking about an additional edge case like this for example `https://camrose.laserfiche.com/LFRepositoryAPI/test/testsefsdfsd` which is why i hardcoded that

File: src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java (hunk 8 above)
Comment (lf-jason-jia) [reply src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java]: I think Ray is working on something so all cloud dev is not working right now. I am working on getting a new environment on cloud test but it's not done.

File: src/main/java/com/laserfiche/api/client/model/ProblemDetails.java (hunk 16 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/model/ProblemDetails.java:9]: @lf-jason-jia which specific method are you referring to? since the repo client version is auto-generated so i just removed both the generated and schema annotations, along with all the non getter/setter methods

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java (hunk 12 above)
Comment (lf-jason-jia) [reply src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java:1]: If this is the case, shall we follow existing class naming? For the client credential flow, we call the handler: OAuthClientCredentialHandler. Since it's just password flow, shall we just call it OAuthPasswordHandler?

File: src/test/java/com/laserfiche/api/client/integration/UsernamePasswordHandlerTest.java (hunk 18 above)
Comment (steven-xu-lf) [reply src/test/java/com/laserfiche/api/client/integration/UsernamePasswordHandlerTest.java]: done

File: src/main/java/com/laserfiche/api/client/APIServer/TokenClient.java (hunk 20 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/APIServer/TokenClient.java]: done

File: src/test/java/com/laserfiche/api/client/unit/UsernamePasswordHandlerTest.java (hunk 21 above)
Comment (steven-xu-lf) [reply src/test/java/com/laserfiche/api/client/unit/UsernamePasswordHandlerTest.java]: done

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java (hunk 22 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java]: done

File: src/test/java/com/laserfiche/api/client/unit/UsernamePasswordHandlerTest.java (hunk 19 above)
Comment (lf-jason-jia) [reply src/test/java/com/laserfiche/api/client/unit/UsernamePasswordHandlerTest.java]: OK, sounds good.

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java (hunk 10 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java]: done

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java (hunk 9 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java]: done

File: src/main/java/com/laserfiche/api/client/model/OffsetDateTimeDeserializer.java (hunk 11 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/model/OffsetDateTimeDeserializer.java]: done

File: src/test/java/com/laserfiche/api/client/integration/BaseTest.java (hunk 7 above)
Comment (steven-xu-lf) [reply src/test/java/com/laserfiche/api/client/integration/BaseTest.java]: done

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java (hunk 5 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java]: done

File: src/main/java/com/laserfiche/api/client/APIServer/SessionKeyInfo.java (hunk 1 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/APIServer/SessionKeyInfo.java]: done

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java (hunk 10 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java]: done

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java (hunk 9 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java]: done

File: src/main/java/com/laserfiche/api/client/model/ProblemDetails.java (hunk 16 above)
Comment (lf-jason-jia) [reply src/main/java/com/laserfiche/api/client/model/ProblemDetails.java:9]: Hi @sxu1998 , that's fine. I guess we could decide on that later. I added it as a reminder for future. But I guess I should create a story for those reminders.

File: src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java (hunk 6 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java]: updated to new version yesterday

File: src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java (hunk 8 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/APIServer/TokenClientImpl.java]: updated to new version yesterday

File: src/main/java/com/laserfiche/api/client/APIServer/TokenClient.java (hunk 3 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/APIServer/TokenClient.java]: done

File: src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java (hunk 12 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/httphandlers/UsernamePasswordHandler.java:1]: changed it back to the original name along with UT and integration test names

File: src/main/java/com/laserfiche/api/client/model/ProblemDetails.java (hunk 16 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/model/ProblemDetails.java:9]: so i guess i will close this issue

File: src/main/java/com/laserfiche/api/client/model/OffsetDateTimeDeserializer.java [hunk 23]
```diff
@@ -0,0 +1,32 @@
+package com.laserfiche.api.client.model;
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/model/OffsetDateTimeDeserializer.java]: Can we put this class in `package com.laserfiche.api.client.deserialization`?

Comment (lf-jason-jia) [reply src/main/java/com/laserfiche/api/client/model/OffsetDateTimeDeserializer.java]: Can we also move `TokenClientObjectMapper` to `package com.laserfiche.api.client.deserialization`?

Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/model/OffsetDateTimeDeserializer.java]: ok done, and i also changed `selfhosted` package name back to `apiserver` since it seems like we agree on that?

File: src/main/java/com/laserfiche/api/client/apiserver/TokenClientImpl.java [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+                .builder()
+                .addModule(module)
+                .disable(MapperFeature.CAN_OVERRIDE_ACCESS_MODIFIERS)
+                .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
+                .enable(MapperFeature.ACCEPT_CASE_INSENSITIVE_ENUMS)
+                .build();
+    }
+
+    @Override
+    public CompletableFuture<SessionKeyInfo> createAccessToken(String repoId, CreateConnectionRequest body) {
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/apiserver/TokenClientImpl.java:45]: Can we give the variable of type `CreateConnectionRequest` a better name? Does `body` convey its true meaning?

File: src/main/java/com/laserfiche/api/client/apiserver/TokenClientImpl.java [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+                new Object[]{repoId});
+        return Unirest
+                .post(baseUrl + "/v1/Repositories/{repoId}/Token")
+                .routeParam(pathParameters)
+                .header("Accept", "application/json")
+                .header("Content-Type", "application/x-www-form-urlencoded")
+                .field("grant_type", "password")
+                .field("username", body
+                        .getUsername()
+                        .toString())
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/apiserver/TokenClientImpl.java]: It's returning `String` so no need for `toString`.

File: src/main/java/com/laserfiche/api/client/apiserver/TokenClientImpl.java [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+                .field("password", body
+                        .getPassword()
+                        .toString())
```
Comment (lf-jason-jia) [inline src/main/java/com/laserfiche/api/client/apiserver/TokenClientImpl.java]: It's returning `String` so no need for `toString`.

File: src/main/java/com/laserfiche/api/client/apiserver/TokenClientImpl.java (hunk 24 above)
Comment (lf-jason-jia) [reply src/main/java/com/laserfiche/api/client/apiserver/TokenClientImpl.java:45]: Also, why do we call it `CreateConnectionRequest`?

File: src/main/java/com/laserfiche/api/client/apiserver/TokenClientImpl.java (hunk 25 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/apiserver/TokenClientImpl.java]: ok done for both

File: src/main/java/com/laserfiche/api/client/apiserver/TokenClientImpl.java (hunk 24 above)
Comment (steven-xu-lf) [reply src/main/java/com/laserfiche/api/client/apiserver/TokenClientImpl.java:45]: oh i just took this object name from the C# client lib as well as the variable name `body`, and i believe the variable name `body` is supposed to represent the request body of that self-hosted token api

Comment (lf-jason-jia) [reply src/main/java/com/laserfiche/api/client/apiserver/TokenClientImpl.java:45]: OK, thanks for expalanation.

**Review by lf-jason-jia — APPROVED** [review]
